ability to bind to non-observable

Manuel-S opened the pull request:
Implemented the way the standard knockout bindings update values, which also updates plain properties if they're not observable.

This is needed to use the binding with knockout-es5 properties.

Manuel-S committed:
ability to bind to non-observable

Implemented the way the standard knockout bindings update values, which also updates plain properties if they're not observable.

This is needed to use the binding with knockout-es5 properties.

cmbankester commented on knockout-tinymce.coffee:
```diff
@@ -7,7 +7,6 @@
     defaults: {}
     extensions: {}
     init: (element, valueAccessor, allBindings, viewModel, bindingContext) ->
-      throw "valueAccessor must be writeable and observable"  unless ko.isWriteableObservable(valueAccessor())
```
We should replace this with an `Error` instance so that the stack trace is available on all browsers

Manuel-S replied:
that line is removed since after this change, the binding can write back to normal properties. Basically, even this should work:
`data-bind="tinymce: prop"`
`var vm = {}; ko.applyBindings(vm);`
and the vm would get a simple string property with the value from tinymce whenever it changes.

cmbankester replied:
(facepalm) It's a slow morning for me.

cmbankester commented:
Aside from the comment I left on `knockout-tinymce.coffee`, it looks good! Fix that up and I'll get it merged in.

Also: thanks for the nod to knockout-es5, I'd never seen that before

cmbankester commented:
I found one issue while testing this: it seems to only work when using the debug version of knockout.js.

Manuel-S commented:
Hm that's not good. My guess is the method is not exported in the min version, might have to copy it.

Manuel-S committed:
pull writeValueToProperty into binding

Since it is not exported (yet? https://github.com/knockout/knockout/pull/2088 ), we use our own version of writeValueToProperty.
We also use _twoWayBindings as it is the exported collection.

Manuel-S commented:
Can you try again with this? Pulled a small utility function into the binding and used the exported member _twoWayBindings.

Manuel-S committed:
slighty improve ui reactiveness

Manuel-S committed:
improve performance

improve the performance by caching the last "write" from the binding to the viewmodel and not trying to update tinymce if it is the same value that was just written.

cmbankester commented on knockout-tinymce.coffee:
```diff
@@ -1,4 +1,5 @@
 (($, ko) ->
+  cache = ""
```
Since the scope of the `cache` variable is outside the scope of the `init` function, only one `cache` variable will be used for all bound tinymce editors. Perhaps we should use a [`WeakMap`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap) with `element` as the keys?

Manuel-S replied:
Right, I was thinking that it wouldn't be a problem, but on closer inspection there is one case I could think of where it could be troublesome. A WeakMap would probably solve this, but requires a polyfill and hence a lot of additional code. A simpler solution would probably be to just have two variables:
cachedContent: string;
cachedInstance: Tinymce.Editor | Element;
the reason for this whole thing was that typing felt slow and the profiler showed that innerHTML was being called a lot, and a lot of those calls came from the getContent() != value check. But it should probably also be set in the update handler...

cmbankester replied:
Yeah, I'm totally on board with adding some caching to reduce the number of `getContent` calls. As far as I can tell, there is no way to pollyfill a `WeakMap`, but we could fall back to a simple `Object` key-value store in cases where `WeakMap` doesn't exist

Manuel-S replied:
As long as we clean it up well it should work, I just think it's overkill.

cmbankester commented on knockout-tinymce.coffee:
```diff
@@ -41,11 +42,19 @@
       return
 
+  writeValueToProperty = (property, allBindings, key, value, checkIfDifferent) -> 
+    if !property or !ko.isObservable property
```
Might be more readable as

``` coffee
unless property and ko.isObservable(property)
```

?

Manuel-S replied:
it's honestly just a copy of the function from the original ko source that is hidden in the minified version.

Manuel-S committed:
also save the last used instance in the binding cache

Manuel-S committed:
fix condition

that's what you'd call an oopsie.

Manuel-S commented:
Do you think this is ready to be merged?

Manuel-S commented:
bump, waiting to integrate this into my project.

Manuel-S commented:
![chuck-norris](https://cloud.githubusercontent.com/assets/1771314/17250364/4e4ac7d6-55a4-11e6-90b1-6cac9750f5c0.gif)
